[Security] make LdapBindAuthenticationProvider capable of searching for the DN

lsmith77 opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #16823, #20905 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/7420

I guess due to the separation between the user and auth provider something like the following isn't ok  (note: the following works just fine for me but if course in the end the username is the DN and not the user provided shorter username):

```diff
diff --git a/src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php b/src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php
index 5ebb09a..18d7825 100644
--- a/src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php
+++ b/src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php
@@ -70,7 +70,7 @@ class LdapBindAuthenticationProvider extends UserAuthenticationProvider
      */
     protected function checkAuthentication(UserInterface $user, UsernamePasswordToken $token)
     {
-        $username = $token->getUsername();
+        $username = $user->getUsername();
         $password = $token->getCredentials();
 
         if ('' === $password) {
@@ -78,10 +78,7 @@ class LdapBindAuthenticationProvider extends UserAuthenticationProvider
         }
 
         try {
-            $username = $this->ldap->escape($username, '', LdapInterface::ESCAPE_DN);
-            $dn = str_replace('{username}', $username, $this->dnString);
-
-            $this->ldap->bind($dn, $password);
+            $this->ldap->bind($username, $password);
         } catch (ConnectionException $e) {
             throw new BadCredentialsException('The presented password is invalid.');
         }
diff --git a/src/Symfony/Component/Security/Core/User/LdapUserProvider.php b/src/Symfony/Component/Security/Core/User/LdapUserProvider.php
index fc42419..8194c4c 100644
--- a/src/Symfony/Component/Security/Core/User/LdapUserProvider.php
+++ b/src/Symfony/Component/Security/Core/User/LdapUserProvider.php
@@ -115,7 +115,7 @@ class LdapUserProvider implements UserProviderInterface
     {
         $password = $this->getPassword($entry);
 
-        return new User($username, $password, $this->defaultRoles);
+        return new User($entry->getDn(), $password, $this->defaultRoles);
     }
 
     /**
```

Therefore I created an entire new auth provider.

lsmith77 committed:
make LdapBindAuthenticationProvider capable of searching for the DN

lsmith77 commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/FormLoginLdapFactory.php:
```diff
@@ -26,15 +26,33 @@ class FormLoginLdapFactory extends FormLoginFactory
+        
+        $provider = 'security.authentication.provider.ldap_'.$method.'.'.$id;
+
+        switch($method) {
```
there is probably a more elegant way to do this without duplicating so much code

lsmith77 commented on src/Symfony/Component/Security/Core/Authentication/Provider/LdapQueryAuthenticationProvider.php:
```diff
@@ -0,0 +1,98 @@
+                throw new BadCredentialsException('The presented password is invalid.');
+            }
+
+            $this->ldap->bind($result->getIterator()->current()->getDn(), $password);
```
this could probably be rewritten as `$this->ldap->bind($result[0]->getDn(), $password);`

lsmith77 commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/FormLoginLdapFactory.php:
```diff
@@ -26,15 +26,33 @@ class FormLoginLdapFactory extends FormLoginFactory
+
+        $provider = 'security.authentication.provider.ldap_'.$method.'.'.$id;
+
+        switch ($method) {
```
there is probably a more elegant way to reduce the code duplication

nicolas-grekas commented:
ping @csarrazi

lsmith77 commented on src/Symfony/Component/Security/Core/Authentication/Provider/LdapQueryAuthenticationProvider.php:
```diff
@@ -0,0 +1,98 @@
+ *
+ * The only way to check user credentials is to first find the DN via query and then bind
+ *
+ * @author Charles Sarrazin <charles@sarraz.in>
```
@csarrazi since I used `LdapBindAuthenticationProvider` as the base I figured I should leave you as an author

lsmith77 replied:
BTW I did however think its not appropriate to extend from `LdapBindAuthenticationProvider` even though it would reduce some code duplication.

lsmith77 replied:
another approach could be to integrate the logic into `LdapBindAuthenticationProvider`

csarrazi replied:
I think it would actually be better to integrate the logic inside `LdapBindAuthenticationProvider`.

lsmith77 replied:
@csarrazi ok .. how should I handle the new parameter without breaking BC on the constructor signature? use setter injection?

lsmith77 replied:
something like

```
diff --git a/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/FormLoginLdapFactory.php b/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/FormLoginLdapFactory.php
index cd4158d..3d9d4b2 100644
--- a/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/FormLoginLdapFactory.php
+++ b/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/FormLoginLdapFactory.php
@@ -27,7 +27,7 @@ class FormLoginLdapFactory extends FormLoginFactory
     protected function createAuthProvider(ContainerBuilder $container, $id, $config, $userProviderId)
     {
         $provider = 'security.authentication.provider.ldap_bind.'.$id;
-        $container
+        $definition = $container
             ->setDefinition($provider, new ChildDefinition('security.authentication.provider.ldap_bind'))
             ->replaceArgument(0, new Reference($userProviderId))
             ->replaceArgument(1, new Reference('security.user_checker.'.$id))
@@ -36,6 +36,10 @@ class FormLoginLdapFactory extends FormLoginFactory
             ->replaceArgument(4, $config['dn_string'])
         ;
 
+        if (!empty($config['query_string'])) {
+            $definition->addMethodCall('setQueryString', [$config['query_string']]);
+        }
+
         return $provider;
     }
 
@@ -47,6 +51,7 @@ class FormLoginLdapFactory extends FormLoginFactory
             ->children()
                 ->scalarNode('service')->defaultValue('ldap')->end()
                 ->scalarNode('dn_string')->defaultValue('{username}')->end()
+                ->scalarNode('query_string')->end()
             ->end()
         ;
     }
diff --git a/src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php b/src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php
index 5ebb09a..3dfe028 100644
--- a/src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php
+++ b/src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php
@@ -33,6 +33,7 @@ class LdapBindAuthenticationProvider extends UserAuthenticationProvider
     private $userProvider;
     private $ldap;
     private $dnString;
+    private $queryString;
 
     /**
      * Constructor.
@@ -54,6 +55,16 @@ class LdapBindAuthenticationProvider extends UserAuthenticationProvider
     }
 
     /**
+     * Set a query string to use in order to find a DN for the username
+     *
+     * @param string $queryString
+     */
+    public function setQueryString($queryString)
+    {
+        $this->queryString = $queryString;
+    }
+
+    /**
      * {@inheritdoc}
      */
     protected function retrieveUser($username, UsernamePasswordToken $token)
@@ -79,7 +90,21 @@ class LdapBindAuthenticationProvider extends UserAuthenticationProvider
 
         try {
             $username = $this->ldap->escape($username, '', LdapInterface::ESCAPE_DN);
-            $dn = str_replace('{username}', $username, $this->dnString);
+
+            if ($this->queryString) {
+                $query = str_replace('{username}', $username, $this->queryString);
+
+                $query = $this->ldap->query($this->dnString, $query);
+                $result = $query->execute();
+                if (1 !== $result->count()) {
+                    throw new BadCredentialsException('The presented username is invalid.');
+                }
+
+                $dn = $result[0]->getDn();
+            } else {
+                $username = $this->ldap->escape($username, '', LdapInterface::ESCAPE_DN);
+                $dn = str_replace('{username}', $username, $this->dnString);
+            }
 
             $this->ldap->bind($dn, $password);
         } catch (ConnectionException $e) {
```

csarrazi replied:
Yup, setter injection would be the way to go. By default the new attribute should be null, and the current behaviour should be the correct one in this case.

csarrazi replied:
One thing, though. It seems that you have duplicate logic in the `else` part of your `if` statement:

```php
$username = $this->ldap->escape($username, '', LdapInterface::ESCAPE_DN);
```

lsmith77 commented:
@csarrazi pushed (and fixed the duplicate code)

lsmith77 commented:
added some tests

lsmith77 commented:
I will test it tomorrow again when I have direct access to the LDAP server of the project where I am implementing this to confirm the tests are sufficient and the code is indeed working.

nietonfir commented:
Just tested  (both the original PR and the updated one - against current [stable@3.2.2](https://github.com/symfony/symfony-standard/commit/d68e2e0819e1443d53f0a1a10655c21e03e43ba0)) this against an OpenLDAP directory configured with a meta backend serving as a proxy for two different OpenLDAP instances  (thus two different DNs - see #16823), works like a charm.

Your changes can also be applied directly to `HttpBasicLdapFactory`:
```diff
diff --git a/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/HttpBasicLdapFactory.php b/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/HttpBasicLdapFactory.php
index 961af71..4665c07 100644
--- a/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/HttpBasicLdapFactory.php
+++ b/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/HttpBasicLdapFactory.php
@@ -28,7 +28,7 @@ class HttpBasicLdapFactory extends HttpBasicFactory
     public function create(ContainerBuilder $container, $id, $config, $userProvider, $defaultEntryPoint)
     {
         $provider = 'security.authentication.provider.ldap_bind.'.$id;
-        $container
+        $definition = $container
             ->setDefinition($provider, new ChildDefinition('security.authentication.provider.ldap_bind'))
             ->replaceArgument(0, new Reference($userProvider))
             ->replaceArgument(1, new Reference('security.user_checker.'.$id))
@@ -40,6 +40,11 @@ class HttpBasicLdapFactory extends HttpBasicFactory
         // entry point
         $entryPointId = $this->createEntryPoint($container, $id, $config, $defaultEntryPoint);
 
+
+        if (!empty($config['query_string'])) {
+            $definition->addMethodCall('setQueryString', array($config['query_string']));
+        }
+
         // listener
         $listenerId = 'security.authentication.listener.basic.'.$id;
         $listener = $container->setDefinition($listenerId, new ChildDefinition('security.authentication.listener.basic'));
@@ -57,6 +62,7 @@ class HttpBasicLdapFactory extends HttpBasicFactory
             ->children()
                 ->scalarNode('service')->defaultValue('ldap')->end()
                 ->scalarNode('dn_string')->defaultValue('{username}')->end()
+                ->scalarNode('query_string')->end()
             ->end()
         ;
     }
```

lsmith77 commented:
@nietonfir can you send a PR to this PR to apply the above changes?

Commit:
Update HttpBasicLdapFactory

so that it matches the current FormLoginLdapFactory ldap implementation,
where the DN can be search for before binding.

nietonfir commented:
@lsmith77 created a PR to your branch. Hope that's what you meant… ;-)

lsmith77 committed:
Merge pull request #1 from nietonfir/http_basic_ldap

Update HttpBasicLdapFactory

lsmith77 commented:
thx

lsmith77 commented:
added support for HTTP basic as well and linked a doc PR.

will do final testing on my end tomorrow.

@csarrazi are the unit tests sufficient or do we also need some functional tests?

csarrazi commented:
I don't think we need functional tests for this one. Functional tests are limited to the Ldap component's integration with OpenLDAP. The component was not updated here, so no functional tests are needed. Unit tests are sufficient here.

lsmith77 commented:
@csarrazi ok fixed your comments and tested it against the production LDAP server. all is well to merge from my side

lsmith77 commented on src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php:
```diff
@@ -79,7 +90,20 @@ protected function checkAuthentication(UserInterface $user, UsernamePasswordToke
+                $query = str_replace('{username}', $username, $this->queryString);
+
+                $query = $this->ldap->query($this->dnString, $query);
+                $result = $query->execute();
```
done

lsmith77 commented on src/Symfony/Component/Security/Core/Authentication/Provider/LdapBindAuthenticationProvider.php:
```diff
@@ -79,7 +90,20 @@ protected function checkAuthentication(UserInterface $user, UsernamePasswordToke
+
+            if ($this->queryString) {
+                $query = str_replace('{username}', $username, $this->queryString);
+
```
done

csarrazi commented:
:+1:

fabpot commented:
Thank you @lsmith77.
